### ISSUE TITLE
NUMA Multi-Process optimization

### ISSUE THREAD

**Issue by msaroufim** [pytorch/serve#1198]
## Is your feature request related to a problem? Please describe.
The goal of this effort is to figure out how to increase CPU throughput by applying ideas from this blog post to torchserve https://huggingface.co/blog/bert-cpu-scaling-part-1

Historically we've seen issues with torchserve CPU performance that were solved either by changing the `torch_set_num_threads(1)` #1138 or by setting `workers=1` and increasing the batch size #1167 but both of these were not actually taking advantage of the number of cores and so the lesson to actually use CPUs with smaller core counts because the performance/price tradeoff wasn't justifiable. But this story was failing to take `numcactl` into account.

`numactl` essentially guarantees an affinity between program and processors so multiple Pytorch programs won't be competing for all resources and you can actually get linear speedups. 

So great, how does this work with torchserve?

## Describe the solution
As of now it's not possible to run multiple instances of torchserve on the same machine to get linear speedups in the number of processors and numa nodes https://github.com/pytorch/serve/issues/1009

But as a workaround it's possible to create multiple Docker instances each with different port to run multiple torchserve instances on the same machine.
First docker instance

```
numactl -C 0-23 -m 0 docker run --rm --shm-size=1g \
        --ulimit memlock=-1 \
        --ulimit stack=67108864 \
        -p8080:8080 \
        -p8081:8081 \
        -p8082:8082 \
        -p7070:7070 \
        -p7071:7071 \
```

And then the second docker instance

```
numactl -C 24-47 -m 1 docker run --rm --shm-size=1g \
        --ulimit memlock=-1 \
        --ulimit stack=67108864 \
        -p8083:8083 \
        -p8084:8084 \
        -p8084:8084 \
        -p7072:7072 \
        -p7072:7072 \
```
Benchmarking these should not show any performance improvements but since there will effectively be 2 torchserve instances then they won't be contending with each other the throughput is effectively doubled.

It also doesn't have to stop at just 2 instances of torchserve, ideally the setup would be the benchmark the least amount of CPU cores needed and only allocate those

`numactl -C 0-11 -m 0` would allocate cores 0 to 11 from numa node 1

## Sanity Check

1. To make sure it all works, we should use the Apache Benchmark suite and see what's the smallest number of cores we can go down to for a BERT model without seeing a performance degradation and then start the parallel runs to confirm. 

2. It's also important to vary the `torch_set_num_threads()` to confirm whether it influences performance when used with `numactl`

3. And as an upper bound or baseline we can also try using the IPEX Intel optimized image with torchserve docker https://github.com/ashokei/serve/pull/1

## Potential risks

The blog post assumes that the code is running on a bare metal instance https://huggingface.co/blog/bert-cpu-scaling-part-1 to avoid any kind of cloud based virtualization. Unfortunately, docker introduces the same of kind of virtualization so even if NUMA is supported with Pytorch it may require more work to make multiple instances of torchserve work on the same machine.

Interestingly enough @HamidShojanazeri is pointing out that a native torchserve solution that is `numactl` may be better. So with that in mind, we can use the approach above as a POC since it doesn't require any changes to torchserve before we invest in a real solution.

cc: @HamidShojanazeri @chauhang @nskool

**Comment by msaroufim** [comment]
Alternatively a simpler solution is to use cpuset in docker https://unix.stackexchange.com/questions/312382/how-to-achieve-processor-affinity-in-containers

**Comment by msaroufim** [comment]
It seems like 1-2ms latency with BERT is possible - how do we get there https://twitter.com/Thom_Wolf/status/1436308987725680640?s=20

**Comment by nskool** [comment]
Was able to reproduce the results you shared @msaroufim for the batch size of 8, using cpuset. As you said, use of numactl/cpuset without virtualization is currently not possible with torchserve, as one may only run a single instance of the server on a particular host.

As mentioned, there seem to be two avenues of tuning without significantly amending Torchserve: (1). specific optimizations for PyTorch using libraries such as TensorRT, IPEX. (2). possibility of minor improvements by fine-tuning the LargePages configuration on the host, or using the jvm process.

Attaching result of some experiments with two configurations: TransparentHugePages, and HugeTLB pages. There are two ways of tuning these values. (a). Modifying the config on the host (requires elevated privileges), (b). Using jvm args supplied with TS config.

* Primary results show a slightly significant improvement when using a, and a marginal improvement when using b. Moreover, the improvement is more effective in reducing latency than improving throughput.

 | Benchmark |Model |Concurrency |Requests |TS failed requests |TS throughput |TS latency P50 |TS latency P90 |TS latency P99 |TS latency mean |TS error rate |Model_p50 |Model_p90 |Model_p99 |predict_mean |handler_time_mean |waiting_time_mean |worker_thread_mean |
 |--- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
 | With THP disabled | bert | 100 | 10000 | 0 | 22.00 | 4526 | 4948 | 5358 | 4546.040 | 0.0 | 1319.73 | 1325.12 | 1326.79 | 1453.93 | 1453.86 | 3061.93 | 1.33 |
 | With THP disabled | bert | 100 | 10000 | 0 | 22.09 | 4472 | 4950 | 5449 | 4526.681 | 0.0 | 1349.25 | 1353.92 | 1354.82 | 1447.08 | 1447.01 | 3048.11 | 1.26 |
 | With THP enabled | bert | 100 | 10000 | 0 | 23.14 | 4199 | 4680 | 5212 | 4321.051 | 0.0 | 1341.2 | 1344.94 | 1345.87 | 1382.52 | 1382.45 | 2911.4 | 1.08 |
 | With THP enabled | bert | 100 | 10000 | 0 | 23.01 | 4271 | 4758 | 5199 | 4363.243 | 0.0 | 1322.36 | 1331.09 | 1332.77 | 1394.26 | 1394.2 | 2937.51 | 1.1 |
 | With +UseLargePages JVM enabled | bert | 100 | 10000 | 0 | 21.98 | 4517 | 5004 | 5495 | 4550.391 | 0.0 | 1308.99 | 1314.38 | 1316.08 | 1455.0 | 1454.93 | 3064.66 | 1.28 |
 | With +UseTransparentHugePages JVM enabled | bert | 100 | 10000 | 0 | 22.24 | 4437 | 4913 | 5391 | 4495.538 | 0.0 | 1335.48 | 1344.17 | 1345.22 | 1437.91 | 1437.84 | 3028.36 | 1.09 |
 | With +UseLargePages LargePageSizeInBytes=256m JVM enabled | bert | 100 | 10000 | 0 | 22.17 | 4454 | 4937 | 5424 | 4511.475 | 0.0 | 1338.83 | 1343.78 | 1345.49 | 1442.13 | 1442.06 | 3037.99 | 1.24 |

---

`On HOST`
Command to enable THP (TransparentHugePages):
`sudo -S sh -c "echo always > /sys/kernel/mm/transparent_hugepage/enabled"`

Command to set THP to default value i.e. madvise:
 `sudo -S sh -c "echo madvise > /sys/kernel/mm/transparent_hugepage/enabled"`

*or*

`Using JVM args` in TS config file:
Add the following to vmargs to use HugeTLB pages
`-XX:+UseLargePages XX:LargePageSizeInBytes=256m`

Add the following to vmargs to use TransparentHugePages:
`-XX:+UseTransparentHugePages`

**Comment by msaroufim** [comment]
So next question may be - why can we only run one torchserve instance at a time given that we can remap ports with `config.properties`

It's possible that this is an artificial constraint, so one way to investigate would be to skip these lines and see what happens

Backend: https://github.com/pytorch/serve/blob/master/ts/model_server.py#L49
Frontend: https://github.com/pytorch/serve/blob/master/frontend/tools/gradle/launcher.gradle#L15

**Comment by msaroufim** [comment]
IPEX is all you need

managed to shave down CPU inference for BERT all the way down to 36ms if using m6i.4x!!
